samza-core/src/main/java/org/apache/samza/util/ThreadUtil.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.util;
+
+import java.lang.management.LockInfo;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MonitorInfo;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ThreadUtil {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ThreadUtil.class);
+  private static final ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
+
+  public static void logThreadDump(String message) {
+    try {
+      ThreadInfo[] threadInfo = THREAD_MX_BEAN.dumpAllThreads(true, true);
+      StringBuilder sb = new StringBuilder();
+      sb.append(message).append("\n");
+      for (ThreadInfo ti: threadInfo) {
+        sb.append(toString(ti)).append("\n");
+      }
+      System.out.println(sb.toString());
+    } catch (Exception e) {
+      LOGGER.error("Could not get and log a thread dump.", e);
+    }
+  }
+
+  /**
+   * Copy of ThreadInfo#toString() without the hardcoded MAX_FRAMES = 8 restriction on thread stack depth.
+   *
+   * Returns a string representation of this thread info.
+   * The format of this string depends on the implementation.
+   * The returned string will typically include
+   * the thread name, the getThreadId thread ID, its state,
+   * and a stack trace if any.
+   *
+   * @return a string representation of this thread info.
+   */
+  private static String toString(ThreadInfo info) {
+    StringBuilder sb = new StringBuilder("\"" + info.getThreadName() + "\""
+        + " Id=" + info.getThreadId() + " " + info.getThreadState());
+    if (info.getLockName() != null) {
+      sb.append(" on " + info.getLockName());
+    }
+    if (info.getLockOwnerName() != null) {
+      sb.append(" owned by \"" + info.getLockOwnerName() + "\" Id="
+          + info.getLockOwnerId());
+    }
+    if (info.isSuspended()) {
+      sb.append(" (suspended)");
+    }
+    if (info.isInNative()) {
+      sb.append(" (in native)");
+    }
+    sb.append('\n');
+    int i = 0;
+    for (; i < info.getStackTrace().length; i++) {
```

Comment:
Is there any risk in having too large of a log message? Such as if you try to send this to StreamAppender?

Reply:
It's possible. Within LinkedIn we use the open source li-apache-kafka-clients with large message support. This might be an issue for a regular Kafka producer. Any suggestions for a workaround? We can print to stdout instead if you prefer that.

Reply:
One thing that comes to mind is to use a large but finite MAX_FRAMES. But doing that estimate will be kind of arbitrary. Using stdout would work, but it's probably nice if we could do the same logging as everything else, since then users won't have to know to look in a different place.
Although, maybe we don't have to worry about this. I think if StreamAppender send fails, the app still will end up writing to the local log, and since drop.producer.errors is true, the producer will just reset itself. We would miss the log on kafka, but maybe that is ok. Is that understanding correct?

Reply:
That's right. `drop.producer.errors` is enabled for StreamAppender, so this (and possibly other log messages in the batch) will get dropped if the batch is too large. It will still write to local.